# 1.a

In [2]:
import numpy as np

In [4]:
def f(x):
    return (x-2)**2

### Left Riemann Sum

In [20]:
def lrs(f, a, b, n):
    d = (b-a)/n
    return d * sum(f(a + i *d) for i in range(n))

### Right Riemann Sum

In [22]:
def rrs(f, a, b, n):
    d = (b-a)/n
    return d * sum(f(a + i*d) for i in range(1, n+1))

### Trapezoidal Rule

$$ \int^b_af(x) dx \approx \frac{b-a}{2n}\left[f(a)+2\sum^{n-1}_{i=1}{f(x_i)}+f(b)\right] $$

In [24]:
def trapezo(f, a, b, n):
    d = (b-a)/n
    x_mid = np.linspace( a+d, b-d, n-1)
    total = (f(a) + 2*np.sum(f(x_mid)) + f(b)) * 0.5
    return total*d

### Simpson's Rule

$$ \int^b_af(x) dx \approx \frac{b-a}{3n}\left[f(a)+4\sum^{n-1}_{i=1,odd}{f(x_i)}+2\sum^{n-1}_{i=2,even}{f(x_i)}+f(b)\right] $$

In [65]:
def simpson(f, a, b, n):
    if n%2 != 0:
        raise ValueError('辛普森法的 n 需為偶數')
    d = (b-a)/n
    x = np.linspace(a, b, n+1) 
    total = f(a)+f(b)
    total += np.sum(f(x[1:n:2])) * 4
    total += np.sum(f(x[2:n-1:2])) * 2

    return total * d / 3

In [66]:
if __name__ == "__main__":
    a, b = 0, 4
    n = 4
    
    left_res = lrs(f, a, b, n)
    print("Left Riemann Sum =",left_res)
    right_res = rrs(f, a, b, n)
    print("Right Riemann Sum =",right_res)
    trapezo_res = trapezo(f, a, b, n)
    print("Trapezoidal Rule =",trapezo_res)
    sim_res = simpson(f, a, b, n)
    print("Simpson's Rule =",sim_res)

Left Riemann Sum = 6.0
Right Riemann Sum = 6.0
Trapezoidal Rule = 6.0
Simpson's Rule = 5.333333333333333


In [44]:
if __name__ == "__main__":
    a, b = 0, 4
    n = 8 
    
    left_res = lrs(f, a, b, n)
    print("Left Riemann Sum =",left_res)
    right_res = rrs(f, a, b, n)
    print("Right Riemann Sum =",right_res)
    trapezo_res = trapezo(f, a, b, n)
    print("Trapezoidal Rule =",trapezo_res)
    sim_res = simpson(f, a, b, n)
    print("Simpson's Rule =",sim_res)

Left Riemann Sum = 5.5
Right Riemann Sum = 5.5
Trapezoidal Rule = 5.5
Simpson's Rule = 5.333333333333333


In [45]:
if __name__ == "__main__":
    a, b = 0, 4
    n = 16 
    
    left_res = lrs(f, a, b, n)
    print("Left Riemann Sum =",left_res)
    right_res = rrs(f, a, b, n)
    print("Right Riemann Sum =",right_res)
    trapezo_res = trapezo(f, a, b, n)
    print("Trapezoidal Rule =",trapezo_res)
    sim_res = simpson(f, a, b, n)
    print("Simpson's Rule =",sim_res)

Left Riemann Sum = 5.375
Right Riemann Sum = 5.375
Trapezoidal Rule = 5.375
Simpson's Rule = 5.333333333333333


In [46]:
if __name__ == "__main__":
    a, b = 0, 4
    n = 32 
    
    left_res = lrs(f, a, b, n)
    print("Left Riemann Sum =",left_res)
    right_res = rrs(f, a, b, n)
    print("Right Riemann Sum =",right_res)
    trapezo_res = trapezo(f, a, b, n)
    print("Trapezoidal Rule =",trapezo_res)
    sim_res = simpson(f, a, b, n)
    print("Simpson's Rule =",sim_res)

Left Riemann Sum = 5.34375
Right Riemann Sum = 5.34375
Trapezoidal Rule = 5.34375
Simpson's Rule = 5.333333333333333


$$ \int^4_0(x-2)^2 dx  = \frac{1}{3}(x-2)^3\Big|^4_0 =\frac{1}{3}\left[2^3-(-2)^3\right] \approx 5.3333 $$

|  n  |\(I_{L_n}\)|         |\(I_{R_n}\)|         |\(I_{T_n}\)|         |\(I_{S_n}\)|         |
|-----|-----------|---------|-----------|---------|-----------|---------|-----------|---------|
|     | **error** |**ratio**| **error** |**ratio**| **error** |**ratio**| **error** |**ratio**|
|  4  | 0.6777    |         | 0.6777    |         | 0.6777    |         | 0         |         |
|  8  | 0.1667    | 0.246   | 0.1667    | 0.246   | 0.1667    | 0.246   | 0         |**None** |
| 16  | 0.0423    | 0.2537  | 0.0423    | 0.2537  | 0.0423    | 0.2537  | 0         |**None** |
| 32  | 0.0104    | 0.2459  | 0.0104    | 0.2459  | 0.0104    | 0.2459  | 0         |**None** |


# 1.b

給定三點 $u, v, w$，其中：$$u<v<w, v-u=h, w-v=h$$
定義：$$p(x)=f(u)\frac{(x-v)(x-w)}{(u-v)(u-w)}+f(v)\frac{(x-u)(x-w)}{(v-u)(v-w)}+f(w)\frac{(x-u)(x-v)}{(w-u)(w-v)}$$
令：$$v=0, u=-h, w=h$$
可得：
$$p(x)=f(-h)\frac{x(x-h)}{(-h)(-h-h)}+f(0)\frac{(x+h)(x-h)}{(h)(-h)}+f(h)\frac{(x+h)(x)}{(h+h)(h)}$$
$$p(x)=f(-h)\frac{x(x-h)}{2h^2}+f(0)\frac{(x+h)(x-h)}{-h^2}+f(h)\frac{(x+h)(x)}{2h^2}$$
計算：
$$I \overset{\text{Let}}{=}\int_u^wp(x)dx=\int_{-h}^hp(x)dx$$
1. 第一項 
$$\int_{-h}^hf(-h)\frac{x(x-h)}{2h^2}dx=\frac{f(-h)}{2h^2}\int_{-h}^h{x(x-h)}dx=\frac{f(-h)}{2h^2}(\frac{2h^2}{3})=\frac{h}{3}f(-h)$$
2. 第二項
$$\int_{-h}^hf(0)\frac{(x+h)(x-h)}{-h^2}dx=-\frac{f(0)}{h^2}\int_{-h}^h(x+h)(x-h)dx=-\frac{f(0)}{h^2}(-\frac{4h^3}{3})=\frac{4h}{3}f(0)$$
3. 第三項
$$\int_{-h}^hf(h)\frac{(x+h)(x)}{2h^2}dx=-\frac{f(h)}{2h^2}\int_{-h}^h(x+h)(x)dx=\frac{f(h)}{2h^2}(\frac{2h^3}{3})=\frac{h}{3}f(h)$$
整理：
$$I=\frac{h}{3}f(-h)+\frac{4h}{3}f(0)+\frac{h}{3}f(h)$$
$$I=\frac{h}{3}\left(f(-h)+4f(0)+f(h)\right)$$
所以：
$$\int_u^wp(x)dx=\frac{h}{3}\left(f(-h)+4f(0)+f(h)\right)$$
$$\int_u^wp(x)dx=\frac{h}{3}\left(f(u)+4f(v)+f(w)\right)$$

# 1.c

給定 $ a=x_0<x_1<...<x_n=b$，間隔 $d=\frac{b-a}{n}$
梯形法則的近似積分為：
$$I_T=\sum^{n-1}_{i=0}\frac{d}{2}\left({f(x_i)}+f(x_{i+1})\right)$$
實際的積分為：
$$I=\int_a^bf(x)dx$$
---
定義：
$$ p(x) = f(x_i) + (x-x_i) \frac{f(x_{i+1})-f(x_i)}{x_{i+1}-x_i}$$
在 $[x_i, x_{i+1}]$ 中，對 $f(x)$ 做泰勒展開，可得：
$$ f(x) = f(x_i) + (x-x_i)f'(x_i) + \frac{(x-x_i)^2}{2}f"(\xi) $$
所以：
$$ p(x)-f(x) = \left[f(x_i) + (x-x_i) \frac{f(x_{i+1})-f(x_i)}{x_{i+1}-x_i}\right] - \left[f(x_i) + (x-x_i)f'(x_i) + \frac{(x-x_i)^2}{2}f''(\xi)\right]$$
$$ p(x)-f(x) = (x-x_i)\left[\frac{f(x_{i+1})-f(x_i)}{h}\right] - \frac{(x-x_i)^2}{2}f''(\xi)$$
接著對 $f(x_{i+1})$ 做泰勒展開，可得：
$$ f(x_{i+1}) = f(x_i) + (x_{i+1}-x_i)f'(x_i) + \frac{(x_{i+1}-x_i)^2}{2}f''(\eta)$$
$$ f(x_{i+1}) = f(x_i) + hf'(x_i) + \frac{h^2}{2}f''(\eta)$$
移項，同除以 $h$，可得：
$$ \frac{f(x_{i+1}) - f(x_i)}{h} = f'(x_i) + \frac{h}{2}f''(\eta)$$
代回 $p(x)-f(x)$，可得：
$$ p(x) - f(x) = (x-x_i)\frac{h}{2}f''(\eta) - \frac{(x-x_i)^2}{2}f''(\xi)$$
計算區間 $[x_i, x_{i+1}]$ 的誤差積分：
$$ e_i = \int_{x_i}^{x_{i+1}} \left(p(x)-f(x)\right)dx = \int_{x_i}^{x_{i+1}} \left((x-x_i)\frac{h}{2}f''(\eta) - \frac{(x-x_i)^2}{2}f''(\xi)\right)dx$$
1. 第一項
$$ \int_{x_i}^{x_{i+1}}(x-x_i)\frac{h}{2}f''(\eta)dx = \frac{h}{2}f''(\eta)\int_{x_i}^{x_{i+1}}(x-x_i)dx = \frac{h}{2}f''(\eta)\frac{h^2}{2} = \frac{h^3}{4}f''(\eta)$$
2. 第二項
$$\int_{x_i}^{x_{i+1}}\frac{(x-x_i)^2}{2}f''(\xi) = \frac{f''(\xi)}{2}\int_{x_i}^{x_{i+1}}(x-x_i)^2dx =\frac{f''(\xi)}{2} \frac{h^3}{3}= \frac{h^3}{6}f''(\xi)$$

整理：
$$ \int_{x_i}^{x_{i+1}} \left(p(x)-f(x)\right)dx =  \frac{h^3}{4}f''(\eta) - \frac{h^3}{6}f''(\xi) = \frac{h^3}{12}\left(3f''(\eta)-2f''(\xi)\right)$$
所以，誤差總和為：
$$ e = \sum_{i=0}^{n-1}e_i = \sum_{i=0}^{n-1}\frac{h^3}{12}\left(3f''(\eta)-2f''(\xi)\right) = n\frac{h^3}{12}\left(3f''(\eta)-2f''(\xi)\right)=\frac{3f''(\eta)-2f''(\xi)}{12}(b-a)^3\frac{1}{n^2}$$
因為 $f(x)$ 是連續函數，且在 $[a, b]$ 存在上界，故存在一常數 $K$ 使得：
$$ |f''(x)| \leq K \ \  \forall x \in [a, b]$$
故：
$$ |e| \leq \frac{3K+2K}{12}(b-a)^3\frac{1}{n^2} = \frac{5K}{12}(b-a)^3\frac{1}{n^2}$$
得到：
$$ |e| = O(n^{-2})$$
即：
$$ I_T - I = O(n^{-2})$$

# 1.d

In [1]:
import numpy as np

a, b = 0, 4 
N = 100000  

x_random = np.random.uniform(a, b, N)

f_values = (x_random - 2) ** 2

integral_estimate = (b - a) * np.mean(f_values)
round(integral_estimate, 6)

5.331633